File: sdk/c/src/providers/tempmailg.c
/**
 * tempmailg.com：GET /public/{locale} + POST /public/get_messages
 */
#include "tempmail_internal.h"
#include <curl/curl.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#ifdef _MSC_VER
#define tm_strtok_r strtok_s
#else
#define tm_strtok_r strtok_r
#endif

#ifdef _WIN32
#define tm_strcasecmp _stricmp
#else
#include <strings.h>
#define tm_strcasecmp strcasecmp
#endif

#define TMG_ORIGIN "https://tempmailg.com"
#define TMG_TOK_PREFIX "tmg1:"

#define TMG_MAX_COOKIES 48
#define TMG_MAX_COOKIE_VAL 4096

typedef struct {
    char key[96];
    char val[512];
} tm_gcookie_t;

static void tm_g_locale(const char *domain, char *out, size_t cap) {
    if (!domain || !domain[0]) {
        snprintf(out, cap, "zh");
        return;
    }
    for (const char *p = domain; *p; p++) {
        if (*p == '/' || *p == '?' || *p == '#' || *p == '\\') {
            snprintf(out, cap, "zh");
            return;
        }
    }
    snprintf(out, cap, "%s", domain);
}

static int tm_g_parse_cookie_hdr(const char *hdr, tm_gcookie_t *tab, int *n) {
    *n = 0;
    if (!hdr || !hdr[0]) return 0;
    char buf[TMG_MAX_COOKIE_VAL];
    size_t hl = strlen(hdr);
    if (hl >= sizeof(buf)) return -1;
    memcpy(buf, hdr, hl + 1);
    char *save = NULL;
    for (char *tok = tm_strtok_r(buf, ";", &save); tok; tok = tm_strtok_r(NULL, ";", &save)) {
        while (*tok == ' ' || *tok == '\t') tok++;
        char *eq = strchr(tok, '=');
        if (!eq || eq == tok) continue;
        *eq = '\0';
        const char *k = tok;
        const char *v = eq + 1;
        while (*v == ' ' || *v == '\t') v++;
        if (*n >= TMG_MAX_COOKIES) return -1;
        snprintf(tab[*n].key, sizeof(tab[*n].key), "%s", k);
        snprintf(tab[*n].val, sizeof(tab[*n].val), "%s", v);
        (*n)++;
    }
    return 0;
}

static int tm_g_cookie_index(tm_gcookie_t *tab, int n, const char *key) {
    for (int i = 0; i < n; i++) {
        if (tm_strcasecmp(tab[i].key, key) == 0) return i;
    }
    return -1;
}

static void tm_g_merge_cookie_responses(const char *prev, const char *newhdr, char *out, size_t cap) {
    tm_gcookie_t tab[TMG_MAX_COOKIES];
    int n = 0;
    tm_g_parse_cookie_hdr(prev, tab, &n);
    tm_gcookie_t tab2[TMG_MAX_COOKIES];
    int n2 = 0;
    if (newhdr && newhdr[0]) tm_g_parse_cookie_hdr(newhdr, tab2, &n2);
    for (int i = 0; i < n2; i++) {
        int j = tm_g_cookie_index(tab, n, tab2[i].key);
        if (j >= 0) {
            snprintf(tab[j].val, sizeof(tab[j].val), "%s", tab2[i].val);
        } else if (n < TMG_MAX_COOKIES) {
            snprintf(tab[n].key, sizeof(tab[n].key), "%s", tab2[i].key);
            snprintf(tab[n].val, sizeof(tab[n].val), "%s", tab2[i].val);
            n++;
        }
    }
    /* sort by key */
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            if (tm_strcasecmp(tab[i].key, tab[j].key) > 0) {
                tm_gcookie_t t = tab[i];
                tab[i] = tab[j];
                tab[j] = t;
            }
        }
    }
    out[0] = '\0';
    size_t pos = 0;
    for (int i = 0; i < n; i++) {
        int add = snprintf(out + pos, cap > pos ? cap - pos : 0, "%s%s=%s",
            (i > 0 && pos > 0) ? "; " : "", tab[i].key, tab[i].val);
        if (add < 0 || (size_t)add >= cap - pos) break;
        pos += (size_t)add;
    }
}

static int tm_g_xsrf_from_hdr(const char *hdr, char *out, size_t cap) {
    tm_gcookie_t tab[TMG_MAX_COOKIES];
    int n = 0;
    tm_g_parse_cookie_hdr(hdr, tab, &n);
    for (int i = 0; i < n; i++) {
        if (tm_strcasecmp(tab[i].key, "XSRF-TOKEN") == 0 || tm_strcasecmp(tab[i].key, "xsrf-token") == 0) {
            snprintf(out, cap, "%s", tab[i].val);
            return 0;
        }
    }
    return -1;
}

static int tm_g_parse_csrf(const char *html, char *out, size_t cap) {
    const char *p = strstr(html, "csrf-token");
    if (!p) return -1;
    p = strstr(p, "content=");
    if (!p) return -1;
    p += 8;
    if (*p == '"' || *p == '\'') p++;
    const char *end = p;
    while (*end && *end != '"' && *end != '\'') end++;
    size_t len = (size_t)(end - p);
    if (len == 0 || len >= cap) return -1;
    memcpy(out, p, len);
    out[len] = '\0';
    return 0;
}

static const char b64tab[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

static char *tm_g_b64_encode(const unsigned char *in, size_t len) {
    size_t olen = 4 * ((len + 2) / 3) + 1;
    char *out = (char *)malloc(olen);
    if (!out) return NULL;
    size_t j = 0;
    for (size_t i = 0; i < len; i += 3) {
        unsigned n = (unsigned)in[i] << 16;
        if (i + 1 < len) n |= (unsigned)in[i + 1] << 8;
        if (i + 2 < len) n |= (unsigned)in[i + 2];
        out[j++] = b64tab[(n >> 18) & 63];
        out[j++] = b64tab[(n >> 12) & 63];
        out[j++] = (i + 1 < len) ? b64tab[(n >> 6) & 63] : '=';
        out[j++] = (i + 2 < len) ? b64tab[n & 63] : '=';
    }
    out[j] = '\0';
    return out;
}

static int tm_g_b64_val(int c) {
    if (c >= 'A' && c <= 'Z') return c - 'A';
    if (c >= 'a' && c <= 'z') return c - 'a' + 26;
    if (c >= '0' && c <= '9') return c - '0' + 52;
    if (c == '+') return 62;
    if (c == '/') return 63;
    return -1;
}

static unsigned char *tm_g_b64_decode(const char *s, size_t *outlen) {
    while (*s && isspace((unsigned char)*s)) s++;
    size_t sl = strlen(s);
    while (sl > 0 && s[sl - 1] == '=') sl--;
    *outlen = (sl * 3) / 4;
    unsigned char *buf = (unsigned char *)malloc(*outlen + 4);
    if (!buf) return NULL;
    size_t j = 0;
    unsigned acc = 0;
    int bits = 0;
    for (size_t i = 0; i < sl; i++) {
        int v = tm_g_b64_val((unsigned char)s[i]);
        if (v < 0) continue;
        acc = (acc << 6) | (unsigned)v;
        bits += 6;
        if (bits >= 8) {
            bits -= 8;
            if (j < *outlen + 4) buf[j++] = (unsigned char)((acc >> bits) & 0xff);
        }
    }
    *outlen = j;
    buf[j] = '\0';
    return buf;
}

static int tm_g_build_page_url(const char *domain, char *url, size_t urlcap) {
    char loc[96];
    tm_g_locale(domain, loc, sizeof(loc));
    CURL *c = curl_easy_init();
    if (!c) return -1;
    char *esc = curl_easy_escape(c, loc, 0);
    curl_easy_cleanup(c);
    if (!esc) return -1;
    int n = snprintf(url, urlcap, TMG_ORIGIN "/public/%s", esc);
    curl_free(esc);
    return (n > 0 && (size_t)n < urlcap) ? 0 : -1;
}

tm_email_info_t *tm_provider_tempmailg_generate(const char *domain) {
    char page_url[512];
    if (tm_g_build_page_url(domain, page_url, sizeof(page_url)) != 0) return NULL;

    char loc[96];
    tm_g_locale(domain, loc, sizeof(loc));

    char refh[640];
    snprintf(refh, sizeof(refh), "Referer: %s", page_url);

    const char *headers_get[] = {
        "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
        "Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        "Accept-Language: zh-CN,zh;q=0.9,en;q=0.8",
        "Cache-Control: no-cache",
        "DNT: 1",
        "Pragma: no-cache",
        refh,
        "Upgrade-Insecure-Requests: 1",
        NULL,
    };

    tm_http_response_t *r1 = tm_http_request(TM_HTTP_GET, page_url, headers_get, NULL, 15);
    if (!r1 || r1->status != 200 || !r1->body) {
        tm_http_response_free(r1);
        return NULL;
    }

    char csrf[512];
    if (tm_g_parse_csrf(r1->body, csrf, sizeof(csrf)) != 0) {
        tm_http_response_free(r1);
        return NULL;
    }

    char cookie_hdr[TMG_MAX_COOKIE_VAL];
    cookie_hdr[0] = '\0';
    if (r1->cookies && r1->cookies[0]) {
        snprintf(cookie_hdr, sizeof(cookie_hdr), "%s", r1->cookies);
    }
    tm_http_response_free(r1);

    char xsrf[512];
    if (tm_g_xsrf_from_hdr(cookie_hdr, xsrf, sizeof(xsrf)) != 0) {
        return NULL;
    }

    cJSON *body1 = cJSON_CreateObject();
    if (!body1) return NULL;
    cJSON_AddStringToObject(body1, "_token", csrf);
    char *post_body = cJSON_PrintUnformatted(body1);
    cJSON_Delete(body1);
    if (!post_body) return NULL;

    char post_url[] = TMG_ORIGIN "/public/get_messages";
    char ckh[4600];
    snprintf(ckh, sizeof(ckh), "Cookie: %s", cookie_hdr);
    char xsh[1200];
    snprintf(xsh, sizeof(xsh), "X-XSRF-TOKEN: %s", xsrf);

    const char *headers_post[] = {
        "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
        "Accept: application/json, text/plain, */*",
        "Accept-Language: zh-CN,zh;q=0.9,en;q=0.8",
        "Content-Type: application/json",
        "Origin: " TMG_ORIGIN,
        refh,
        "Cache-Control: no-cache",
        "Pragma: no-cache",
        "DNT: 1",
        ckh,
        xsh,
        NULL,
    };

    tm_http_response_t *r2 = tm_http_request(TM_HTTP_POST, post_url, headers_post, post_body, 15);
    free(post_body);
    if (!r2 || r2->status != 200 || !r2->body) {
        tm_http_response_free(r2);
        return NULL;
    }

    cJSON *wrap = cJSON_Parse(r2->body);
    if (!wrap) {
        tm_http_response_free(r2);
        return NULL;
    }
    cJSON *st = cJSON_GetObjectItemCaseSensitive(wrap, "status");
    cJSON *mb = cJSON_GetObjectItemCaseSensitive(wrap, "mailbox");
    if (!cJSON_IsTrue(st) || !cJSON_IsString(mb) || !mb->valuestring || !mb->valuestring[0]) {
        cJSON_Delete(wrap);
        tm_http_response_free(r2);
        return NULL;
    }

    char *mailbox = tm_strdup(mb->valuestring);
    if (!mailbox) {
        cJSON_Delete(wrap);
        tm_http_response_free(r2);
        return NULL;
    }

    char merged[TMG_MAX_COOKIE_VAL];
    tm_g_merge_cookie_responses(cookie_hdr, r2->cookies ? r2->cookies : "", merged, sizeof(merged));
    cJSON_Delete(wrap);
    tm_http_response_free(r2);

    cJSON *tokj = cJSON_CreateObject();
    if (!tokj) {
        free(mailbox);
        return NULL;
    }
    cJSON_AddStringToObject(tokj, "l", loc);
    cJSON_AddStringToObject(tokj, "c", merged);
    cJSON_AddStringToObject(tokj, "s", csrf);
    char *jraw = cJSON_PrintUnformatted(tokj);
    cJSON_Delete(tokj);
    if (!jraw) {
        free(mailbox);
        return NULL;
    }

    unsigned char *jb = (unsigned char *)jraw;
    size_t jl = strlen(jraw);
    char *b64 = tm_g_b64_encode(jb, jl);
    free(jraw);
    if (!b64) {
        free(mailbox);
        return NULL;
    }

    size_t tlen = strlen(TMG_TOK_PREFIX) + strlen(b64) + 1;
    char *token = (char *)malloc(tlen);
    if (!token) {
        free(b64);
        free(mailbox);
        return NULL;
    }
    snprintf(token, tlen, "%s%s", TMG_TOK_PREFIX, b64);
    free(b64);

    tm_email_info_t *info = tm_email_info_new();
    if (!info) {
        free(token);
        free(mailbox);
        return NULL;
    }
    info->channel = CHANNEL_TEMPMAILG;
    info->email = mailbox;
    info->token = token;
    return info;
}

tm_email_t *tm_provider_tempmailg_get_emails(const char *token, const char *email, int *count) {
    *count = -1;
    if (!token || strncmp(token, TMG_TOK_PREFIX, strlen(TMG_TOK_PREFIX)) != 0) return NULL;
    const char *b64s = token + strlen(TMG_TOK_PREFIX);
    size_t rawlen = 0;
    unsigned char *raw = tm_g_b64_decode(b64s, &rawlen);
    if (!raw) return NULL;
    cJSON *sess = cJSON_Parse((char *)raw);
    free(raw);
    if (!sess) return NULL;

    const char *loc = "zh";
    cJSON *lj = cJSON_GetObjectItemCaseSensitive(sess, "l");
    if (cJSON_IsString(lj) && lj->valuestring && lj->valuestring[0]) loc = lj->valuestring;
    cJSON *cj = cJSON_GetObjectItemCaseSensitive(sess, "c");
    cJSON *sj = cJSON_GetObjectItemCaseSensitive(sess, "s");
    if (!cJSON_IsString(cj) || !cj->valuestring || !cJSON_IsString(sj) || !sj->valuestring) {
        cJSON_Delete(sess);
        return NULL;
    }
    char *cookie_copy = tm_strdup(cj->valuestring);
    char *csrf_copy = tm_strdup(sj->valuestring);
    cJSON_Delete(sess);
    if (!cookie_copy || !csrf_copy) {
        free(cookie_copy);
        free(csrf_copy);
        return NULL;
    }

    char page_url[512];
    CURL *curl = curl_easy_init();
    if (!curl) {
        free(cookie_copy);
        free(csrf_copy);
        return NULL;
    }
    char *esc = curl_easy_escape(curl, loc, 0);
    curl_easy_cleanup(curl);
    if (!esc) {
        free(cookie_copy);
        free(csrf_copy);
        return NULL;
    }
    snprintf(page_url, sizeof(page_url), TMG_ORIGIN "/public/%s", esc);
    curl_free(esc);

    cJSON *bodyo = cJSON_CreateObject();
    cJSON_AddStringToObject(bodyo, "_token", csrf_copy);
    char *post_body = cJSON_PrintUnformatted(bodyo);
    cJSON_Delete(bodyo);
    if (!post_body) {
        free(cookie_copy);
        free(csrf_copy);
        return NULL;
    }

    char xsrf[512];
    if (tm_g_xsrf_from_hdr(cookie_copy, xsrf, sizeof(xsrf)) != 0) {
        free(post_body);
        free(cookie_copy);
        free(csrf_copy);
        return NULL;
    }

    char refh2[640];
    snprintf(refh2, sizeof(refh2), "Referer: %s", page_url);
    char ckh2[4600];
    snprintf(ckh2, sizeof(ckh2), "Cookie: %s", cookie_copy);
    char xsh2[1200];
    snprintf(xsh2, sizeof(xsh2), "X-XSRF-TOKEN: %s", xsrf);

    char post_url[] = TMG_ORIGIN "/public/get_messages";
    const char *headers_post[] = {
        "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
        "Accept: application/json, text/plain, */*",
        "Accept-Language: zh-CN,zh;q=0.9,en;q=0.8",
        "Content-Type: application/json",
        "Origin: " TMG_ORIGIN,
        refh2,
        "Cache-Control: no-cache",
        "Pragma: no-cache",
        "DNT: 1",
        ckh2,
        xsh2,
        NULL,
    };

    tm_http_response_t *r = tm_http_request(TM_HTTP_POST, post_url, headers_post, post_body, 15);
    free(post_body);
    free(cookie_copy);
    free(csrf_copy);
    if (!r || r->status != 200 || !r->body) {
        tm_http_response_free(r);
        return NULL;
    }

    cJSON *wrap = cJSON_Parse(r->body);
    tm_http_response_free(r);
    if (!wrap) return NULL;
    cJSON *st = cJSON_GetObjectItemCaseSensitive(wrap, "status");
    if (!cJSON_IsTrue(st)) {
        cJSON_Delete(wrap);
        return NULL;
    }
    cJSON *mbx = cJSON_GetObjectItemCaseSensitive(wrap, "mailbox");
    if (cJSON_IsString(mbx) && mbx->valuestring && mbx->valuestring[0]) {
        /* trim compare */
        const char *a = mbx->valuestring;
        const char *b = email;
        while (*a == ' ') a++;
        while (*b == ' ') b++;
        if (tm_strcasecmp(a, b) != 0) {
            cJSON_Delete(wrap);
            return NULL;
        }
    }

    cJSON *msgs = cJSON_GetObjectItemCaseSensitive(wrap, "messages");
    if (!cJSON_IsArray(msgs)) {
        cJSON_Delete(wrap);
        *count = 0;
        return NULL;
    }
    int n = cJSON_GetArraySize(msgs);
    if (n <= 0) {
        cJSON_Delete(wrap);
        *count = 0;
        return NULL;
    }
    tm_email_t *arr = tm_emails_new(n);
    if (!arr) {
        cJSON_Delete(wrap);
        return NULL;
    }
    for (int i = 0; i < n; i++) {
        cJSON *it = cJSON_GetArrayItem(msgs, i);
        if (cJSON_IsObject(it)) {
            tm_email_t ne = tm_normalize_email(it, email);
            arr[i] = ne;
        }
    }
    cJSON_Delete(wrap);
    *count = n;
    return arr;
}

File: sdk/rust/src/providers/tempmailg.rs
/*!
 * tempmailg.com：无 Cookie 罐会话 + POST /public/get_messages
 */

use std::collections::BTreeMap;
use std::sync::LazyLock;

use base64::{engine::general_purpose::STANDARD as B64, Engine as _};
use regex::Regex;
use serde::{Deserialize, Serialize};
use wreq::header::HeaderMap;

use crate::config::{block_on, get_current_ua, http_client_no_cookie_jar};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};

const ORIGIN: &str = "https://tempmailg.com";
const TOK_PREFIX: &str = "tmg1:";

static CSRF_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#"(?i)<meta\s+name="csrf-token"\s+content="([^"]+)""#).expect("csrf re")
});

#[derive(Serialize, Deserialize)]
struct TempmailgSess {
    l: String,
    c: String,
    s: String,
}

fn locale_from_domain(domain: Option<&str>) -> String {
    let s = domain.unwrap_or("").trim();
    if s.is_empty() || s.contains('/') || s.contains('?') || s.contains('#') || s.contains('\\') {
        return "zh".into();
    }
    s.to_string()
}

fn cookie_map(hdr: &str) -> BTreeMap<String, String> {
    let mut m = BTreeMap::new();
    for part in hdr.split(';') {
        let p = part.trim();
        if let Some(i) = p.find('=') {
            let k = p[..i].trim().to_string();
            let v = p[i + 1..].trim().to_string();
            if !k.is_empty() {
                m.insert(k, v);
            }
        }
    }
    m
}

fn cookie_hdr_from_map(m: &BTreeMap<String, String>) -> String {
    m.iter()
        .map(|(k, v)| format!("{}={}", k, v))
        .collect::<Vec<_>>()
        .join("; ")
}

fn merge_set_cookies(prev: &str, headers: &HeaderMap) -> String {
    let mut m = cookie_map(prev);
    for v in headers.get_all("set-cookie") {
        let s = v.to_str().unwrap_or("");
        let nv = s.split(';').next().unwrap_or("").trim();
        if let Some(i) = nv.find('=') {
            let k = nv[..i].trim().to_string();
            let val = nv[i + 1..].trim().to_string();
            if !k.is_empty() {
                m.insert(k, val);
            }
        }
    }
    cookie_hdr_from_map(&m)
}

fn xsrf_from_cookie_hdr(hdr: &str) -> String {
    let m = cookie_map(hdr);
    for key in ["XSRF-TOKEN", "xsrf-token"] {
        if let Some(v) = m.get(key) {
            return v.clone();
        }
    }
    for (k, v) in &m {
        if k.eq_ignore_ascii_case("xsrf-token") {
            return v.clone();
        }
    }
    String::new()
}

fn parse_csrf(html: &str) -> Result<String, String> {
    let cap = CSRF_RE
        .captures(html)
        .ok_or_else(|| "tempmailg: csrf-token not found in page".to_string())?;
    let t = cap.get(1).unwrap().as_str().trim();
    if t.is_empty() {
        return Err("tempmailg: empty csrf-token".into());
    }
    Ok(t.to_string())
}

fn encode_token(sess: &TempmailgSess) -> Result<String, String> {
    let json = serde_json::to_string(sess).map_err(|e| e.to_string())?;
    Ok(format!("{}{}", TOK_PREFIX, B64.encode(json)))
}

fn decode_token(tok: &str) -> Result<TempmailgSess, String> {
    if !tok.starts_with(TOK_PREFIX) {
        return Err("tempmailg: invalid session token".into());
    }
    let raw = B64
        .decode(tok.strip_prefix(TOK_PREFIX).unwrap())
        .map_err(|_| "tempmailg: invalid session token".to_string())?;
    let sess: TempmailgSess =
        serde_json::from_slice(&raw).map_err(|_| "tempmailg: invalid session token".to_string())?;
    if sess.c.is_empty() || sess.s.is_empty() {
        return Err("tempmailg: invalid session token".into());
    }
    Ok(sess)
}

pub fn generate_email(domain: Option<&str>) -> Result<EmailInfo, String> {
    let locale = locale_from_domain(domain);
    let page_url = format!("{ORIGIN}/public/{}", urlencoding::encode(&locale));
    block_on(async {
        let client = http_client_no_cookie_jar();
        let resp = client
            .get(&page_url)
            .header("User-Agent", get_current_ua())
            .header(
                "Accept",
                "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            )
            .header("Accept-Language", "zh-CN,zh;q=0.9,en;q=0.8")
            .header("Cache-Control", "no-cache")
            .header("DNT", "1")
            .header("Pragma", "no-cache")
            .header("Referer", &page_url)
            .header("Upgrade-Insecure-Requests", "1")
            .send()
            .await
            .map_err(|e| format!("tempmailg page: {}", e))?;
        if !resp.status().is_success() {
            return Err(format!("tempmailg page: {}", resp.status()));
        }
        let headers = resp.headers().clone();
        let html = resp.text().await.map_err(|e| e.to_string())?;
        let csrf = parse_csrf(&html)?;
        let mut cookie_hdr = merge_set_cookies("", &headers);
        let xsrf = xsrf_from_cookie_hdr(&cookie_hdr);
        if xsrf.is_empty() {
            return Err("tempmailg: missing XSRF-TOKEN cookie".into());
        }

        let post_url = format!("{ORIGIN}/public/get_messages");
        let body = serde_json::to_string(&serde_json::json!({ "_token": csrf }))
            .map_err(|e| e.to_string())?;
        let resp2 = client
            .post(&post_url)
            .header("User-Agent", get_current_ua())
            .header("Accept", "application/json, text/plain, */*")
            .header("Accept-Language", "zh-CN,zh;q=0.9,en;q=0.8")
            .header("Content-Type", "application/json")
            .header("Origin", ORIGIN)
            .header("Referer", &page_url)
            .header("Cache-Control", "no-cache")
            .header("Pragma", "no-cache")
            .header("DNT", "1")
            .header("Cookie", &cookie_hdr)
            .header("X-XSRF-TOKEN", &xsrf)
            .body(body)
            .send()
            .await
            .map_err(|e| format!("tempmailg get_messages: {}", e))?;
        if !resp2.status().is_success() {
            return Err(format!("tempmailg get_messages: {}", resp2.status()));
        }
        let headers2 = resp2.headers().clone();
        let api_body = resp2.text().await.map_err(|e| e.to_string())?;
        let wrap: serde_json::Value =
            serde_json::from_str(&api_body).map_err(|e| format!("tempmailg json: {}", e))?;
        let ok = wrap.get("status").and_then(|x| x.as_bool()) == Some(true);
        let mailbox = wrap
            .get("mailbox")
            .and_then(|x| x.as_str())
            .unwrap_or("")
            .to_string();
        if !ok || mailbox.is_empty() {
            return Err("tempmailg: create mailbox failed".into());
        }
        cookie_hdr = merge_set_cookies(&cookie_hdr, &headers2);
        let tok = encode_token(&TempmailgSess {
            l: locale,
            c: cookie_hdr,
            s: csrf,
        })?;
        Ok(EmailInfo {
            channel: Channel::Tempmailg,
            email: mailbox,
            token: Some(tok),
            expires_at: None,
            created_at: None,
        })
    })
}

pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let sess = decode_token(token)?;
    let locale = if sess.l.is_empty() {
        "zh".to_string()
    } else {
        sess.l.clone()
    };
    let page_url = format!("{ORIGIN}/public/{}", urlencoding::encode(&locale));
    let post_url = format!("{ORIGIN}/public/get_messages");
    let xsrf = xsrf_from_cookie_hdr(&sess.c);
    let body = serde_json::to_string(&serde_json::json!({ "_token": sess.s })).map_err(|e| e.to_string())?;
    block_on(async {
        let client = http_client_no_cookie_jar();
        let resp = client
            .post(&post_url)
            .header("User-Agent", get_current_ua())
            .header("Accept", "application/json, text/plain, */*")
            .header("Accept-Language", "zh-CN,zh;q=0.9,en;q=0.8")
            .header("Content-Type", "application/json")
            .header("Origin", ORIGIN)
            .header("Referer", &page_url)
            .header("Cache-Control", "no-cache")
            .header("Pragma", "no-cache")
            .header("DNT", "1")
            .header("Cookie", &sess.c)
            .header("X-XSRF-TOKEN", &xsrf)
            .body(body)
            .send()
            .await
            .map_err(|e| format!("tempmailg get_messages: {}", e))?;
        if !resp.status().is_success() {
            return Err(format!("tempmailg get_messages: {}", resp.status()));
        }
        let raw = resp.text().await.map_err(|e| e.to_string())?;
        let wrap: serde_json::Value =
            serde_json::from_str(&raw).map_err(|e| format!("tempmailg json: {}", e))?;
        if wrap.get("status").and_then(|x| x.as_bool()) != Some(true) {
            return Err("tempmailg: get_messages failed".into());
        }
        if let Some(mb) = wrap.get("mailbox").and_then(|x| x.as_str()) {
            if !mb.trim().is_empty() && !mb.trim().eq_ignore_ascii_case(email.trim()) {
                return Err("tempmailg: mailbox mismatch".into());
            }
        }
        let list = wrap
            .get("messages")
            .and_then(|x| x.as_array())
            .cloned()
            .unwrap_or_default();
        let mut out = Vec::with_capacity(list.len());
        for item in list {
            out.push(normalize_email(&item, email));
        }
        Ok(out)
    })
}
